### PR TITLE
[5.4] Use str_random() for generating file names

### PR DESCRIPTION
**Pull request by themsaid**
In reference to https://github.com/laravel/internals/issues/161

### CODE REVIEW

**Comment by taylorotwell** [comment]
Looks like you're missing a namespace import?

**Comment by themsaid** [comment]
Fixed, sorry. Was submitting to 5.3 first and changed branch looks like I dropped this import.

File: src/Illuminate/Http/FileHelpers.php [hunk 1]
```diff
@@ -46,6 +48,6 @@ public function hashName($path = null)
             $path = rtrim($path, '/').'/';
         }
 
-        return $path.md5_file($this->getRealPath()).'.'.$this->guessExtension();
+        return $path.Uuid::uuid4()->toString().'.'.$this->guessExtension();
```
Comment (lucasmichot) [inline src/Illuminate/Http/FileHelpers.php]: @themsaid this is a great new addition.
Nevertheless `md5` is great as it only contains hexadecimals, and allows to structure the filename splitting parts of its name:

`0123456789abcdef.jpg` => `01/23/45/67/89/ab/cd/ef.jpg`

But `UUID4` also contains `-`.

I believe this character could be removed for readability purpose?

```php
str_replace('-', '',  Uuid::uuid4()->toString())...
```

Comment (taylorotwell) [inline src/Illuminate/Http/FileHelpers.php]: It's fine with me to strip dashes.

Comment (vlakoff) [inline src/Illuminate/Http/FileHelpers.php]: Or simply: `bin2hex(random_bytes(16))`.

Comment (jeroennoten) [inline src/Illuminate/Http/FileHelpers.php]: I think a UUID is more likely to be unique than random_bytes, right? UUID's strength is uniqueness, random_bytes's strength is cryptographically secure randomness.

**Comment by a-komarev** [comment]
Isn't `getHex()` get the same strip dashes?
https://github.com/ramsey/uuid/blob/master/src/Uuid.php#L357

**Comment by a-komarev** [comment]
And I have one question about UUIDv4 uniqueness. I had an experience couple of years ago with this package in high load advertisement tracking system. All clicks were marked with UUIDv4, after half a year I found that there were ~3% of ID duplications. After that I'm always trying to use UUIDv5 when I want to make really unique value, but it require to generate namespace UUID first.

File: src/Illuminate/Http/FileHelpers.php (hunk 1 above)
Comment (vlakoff) [inline src/Illuminate/Http/FileHelpers.php]: ramsey/uuid just uses random_bytes() too, see [`RandomBytesGenerator`](https://github.com/ramsey/uuid/blob/master/src/Generator/RandomBytesGenerator.php).

In fact my code above is just the removal of a bunch of classes, factories, generators, etc. the result is the equivalent code without all the winter layers.

**Review by GrahamCampbell — COMMENTED** [review]
> I think a UUID is more likely to be unique than random_bytes, right?

No, basically.

**Comment by vlakoff** [comment]
> or just use str_random(32)

I thought about this, but maybe it's not a good idea because of inconsistencies on case-sensitive vs. case-insensitive filesystems.

**Comment by GrahamCampbell** [comment]
> I thought about this, but maybe it's not a good idea because of inconsistencies on case-sensitive vs. case-insensitive filesystems.

Just lowercase the output of the function?

**Comment by vlakoff** [comment]
The result would be unevenly distributed (bias towards letters), still it would have more entropy than my solution, so it might be better, technically.

Just as a reminder, these kinds of discussions are endless ;)

**Comment by GrahamCampbell** [comment]
Just use str_random, and we're done. No need to f**k about with uuids.

**Comment by taylorotwell** [comment]
Graham does make a good point :smile:

**Comment by vlakoff** [comment]
I had an example in mind for the lowercase need.

If you do path splitting as lucasmichot said above:
0123456789abcdef.jpg => 01/23/45/67/89/ab/cd/ef.jpg

If you reuse some data from windows to linux server, you're very likely to encounter issue.

I think we should really apply a `strtolower`.

**Comment by jeroennoten** [comment]
Just write your own file storing procedure if you need something different, it's not hard. This methods exists really only for convenience.

**Comment by vlakoff** [comment]
Refs the subsequent https://github.com/laravel/framework/commit/5811bc377ef43f5d228b493579cec428e074efe4.
:smile:

**Comment by vlakoff** [comment]
Documentation needs to be updated: https://laravel.com/docs/master/filesystem#file-uploads.

**Pull request merged**